backends/hip-shared/kernels/hip-shared-basis.hip.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed.h>
+#include <hip/hip_runtime.h>
+
+const int sizeMax = 16;
+__constant__ double c_B[sizeMax*sizeMax];
+__constant__ double c_G[sizeMax*sizeMax];
+
+//------------------------------------------------------------------------------
+// Interp device initalization
+//------------------------------------------------------------------------------
+extern "C" int CeedHipInitInterp(CeedScalar *d_B, CeedInt P1d, CeedInt Q1d,
+                                  CeedScalar **c_B_ptr) {
+  const int Bsize = P1d*Q1d*sizeof(CeedScalar);
+  hipMemcpyToSymbol(HIP_SYMBOL(c_B), d_B, Bsize, 0, hipMemcpyDeviceToDevice);
+  hipGetSymbolAddress((void **)c_B_ptr, HIP_SYMBOL(c_B));
+
+  return 0;
+}
+
+//------------------------------------------------------------------------------
+// Grad device initalization
+//------------------------------------------------------------------------------
+extern "C" int CeedHipInitInterpGrad(CeedScalar *d_B, CeedScalar *d_G,
```

Comment:
`CEED_INTERN` should be sufficient here, though if you declare these in a local header, you wouldn't need to spell it out here and the compiler will check that they are compatible. I think this would be better than the manual declaration you currently have in `ceed-hip-shared-basis.c`.

Reply:
I moved these functions to the `ceed-hip-shared.h` header, and marked with `CEED_INTERN` (see 526015bc301544c98d597d81cf7f1f3611bf5ead).
This seemed in line with what other backends do (as we currently do in the MAGMA backend, for example).
I can also make a separate header just for these two functions...if we do that, we might want to do the same for the other backends, though (at least the CUDA backends, to mirror the HIP backends, since they are so closely related)...

Reply:
Ah, cool. It's not necessary to repeat the `extern "C"` part here, so long as it's visible in the header. Looks like there are probably a lot of places where `extern "C"` can be replaced by `CEED_INTERN`, though it needn't be done in this PR. I don't mind it going in the common header at this point.

Reply:
Oh yeah, whoops.  I'll remove the extra `extern "C"`s

Reply:
You're right that we have a lot of `extern "C"`s floating around various backends where we should use `CEED_INTERN`, but for now, I at least changed them in these two new backends - the only `extern "C"`s left should be in the code strings that get fed to `hiprtc`.

Reply:
Thanks!